cli_tools/common/utils/daisyutils/daisy_worker.go

```diff
@@ -74,35 +78,93 @@ func createResourceLabelerIfMissing(env EnvironmentSettings, hooks []interface{}
 }
 
 type defaultDaisyWorker struct {
-	wf     *daisy.Workflow
-	logger logging.Logger
-	env    EnvironmentSettings
-	hooks  []interface{}
+	workflowProvider WorkflowProvider
+	finished         *daisy.Workflow
+	logger           logging.Logger
+	env              EnvironmentSettings
+	hooks            []interface{}
+
+	cancel      chan string
+	cancelGuard sync.Once
 }
 
 // Run runs the daisy workflow with the supplied vars.
-func (w *defaultDaisyWorker) Run(vars map[string]string) error {
-	if err := (&ApplyAndValidateVars{w.env, vars}).PreRunHook(w.wf); err != nil {
-		return err
+func (w *defaultDaisyWorker) Run(vars map[string]string) (err error) {
+	var wf *daisy.Workflow
+	for attempts := 0; attempts <= 1; attempts++ {
+		if wf, err = w.workflowProvider(); err != nil {
```

Comment:
How will this work for e.g. fallback to pd-standard scenario? Wouldn't workflowProvider need to know it's a retry attempt and generate a slighly different workflow for the retry? For that, it would require some kind of an arg (e.g. `attemp`.

Reply:
I added a comment to Tony's question below.